Refactoring. Sparkling drinks support

baby-benz opened the pull request:
Big refactoring for better project scalability. Champagne and sparkling drinks parsing support also has been added.

baby-benz committed:
embed parser application codebase in winestyle-parser-service

NShmendyuk committed:
added cron expression in application properties, check service business on scheduler

baby-benz committed:
large scale project restructuring
change getElementBySmth to selectFirst for performance increase purposes
implement builder pattern by adding additional @Builder and constructor annotations
tell JPA Hibernate to autogenerate table by adding spring.jpa.hibernate.ddl-auto=update to .properties file

baby-benz committed:
Merge remote-tracking branch 'origin/feature/scheduler-upgrade' into dev

# Conflicts:
#	src/main/java/com/wine/to/up/winestyle/parser/service/service/ParserService.java

baby-benz committed:
minor scheduler and entity properties fixes

baby-benz committed:
fix initial condition in parseByPage loop

Commit:
Javadoc added

baby-benz committed:
Merge pull request #1 from Gashad0kuro/dev

Javadoc added

kanootoko committed:
REST API update
Add API to get all wines, wine by id, CSV file with all wines (needs testing), wine with given columns
Change path of API to have prefix "/winestyle"
Change CSVUtility format of output

baby-benz committed:
Merge pull request #2 from baby-benz/kanootoko

REST API update

NShmendyuk committed:
added exception when no entity found by id (no response body yet), added description for scheduler cron

baby-benz committed:
make HttpStatus code and body message more informative
replace custom Exceptions constructors with annotations

baby-benz committed:
add javadoc to getParsedWineWithFields
replace Wine toString() method with ToString annotation

baby-benz committed:
replace ParserService's and WineService's constructor with Constructor annotation
fix getParsedWineWithFields
minor refactoring

baby-benz committed:
make CSVUtility methods invocation through the interface, not directly calling class instance methods.

baby-benz committed:
turn CSVUtility into a full static utility-class using @UtilityClass annotation
fix toCsvFile method header issue

baby-benz committed:
add sparkling wines parser support
global project restructuring
small naming issues fixes

baby-benz committed:
deeper refactoring

baby-benz committed:
ProductPageParser annotation fix

baby-benz committed:
Builder classes bugfix

baby-benz committed:
add handling of one more condition in parseTypeColorSugar and parseColorAndSugar
add sparkling drinks crop year field support

baby-benz committed:
fix parseCropYear title splitting

baby-benz committed:
fix parseCropYear title splitting

baby-benz committed:
final project refactoring
parsing optimizations

baby-benz committed:
final project optimizations
primary codebase for Image scraping

baby-benz committed:
fix Alcohol entity JavaDoc

NShmendyuk commented on src/main/java/com/wine/to/up/winestyle/parser/service/domain/entity/Alcohol.java:
```diff
@@ -0,0 +1,81 @@
+    private String brand;
+
+    @Column
+    private String color;
```
color вроде же не во всех продукциях есть? Надо уточнить

NShmendyuk commented on src/main/java/com/wine/to/up/winestyle/parser/service/domain/entity/Alcohol.java:
```diff
@@ -0,0 +1,81 @@
+
+    @Column(columnDefinition="TEXT")
+    private String description;
+
```
а где color из дегустационных заметок, или он нам не нужен?

NShmendyuk commented on src/main/java/com/wine/to/up/winestyle/parser/service/controller/MainController.java:
```diff
@@ -0,0 +1,102 @@
+        return ResponseEntity.status(HttpStatus.OK).body(parsedAlcohol);
+    }
+
+    @GetMapping("/wine")
```
Лучше /wines писать

NShmendyuk replied:
думаю лучше использовать множественную форму - /wines

NShmendyuk commented on src/main/java/com/wine/to/up/winestyle/parser/service/domain/entity/Alcohol.java:
```diff
@@ -0,0 +1,106 @@
+
+    @Column(columnDefinition="TEXT")
+    private String description;
+
```
а где color из дегустационных заметок, или он нам не нужен?

kanootoko replied:
Вроде, мы его не обрабатываем и не храним

NShmendyuk commented on src/main/java/com/wine/to/up/winestyle/parser/service/service/implementation/controller/ParsingControllerService.java:
```diff
@@ -0,0 +1,83 @@
+            try {
+                parse(SUPPORTED_ALCOHOL_URLS.get("wine"), "wine");
+            } catch (InterruptedException e) {
+                e.printStackTrace();
```
добавь логгирование

NShmendyuk reviewed (commented):
Просмотрел часть, пока что проблема только в шедулере - не запускает два потока для разных типов алкоголя

baby-benz committed:
add database table to collect errors on saving data to main database
update migration sql-file
upgrade parser to process fields not belonginging to HTML-elements

baby-benz committed:
minor changes in REST-endpoints names

baby-benz committed:
fix flyway migration SQL-file

baby-benz committed:
parser no type property bugfix
fix flyway migration SQL-file

baby-benz committed:
fix type property parsing

baby-benz committed:
remove redundant ResponseEntity return type from MainController

kanootoko commented on src/main/java/com/wine/to/up/winestyle/parser/service/controller/MainController.java:
```diff
@@ -0,0 +1,98 @@
+                try {
+                    res.put(fieldName, field.get(parsedAlcohol));
+                } catch (IllegalArgumentException | IllegalAccessException e) {
+                    // TODO: обрабатывать исключения
```
Где-то я это уже видел 🤔
Можно убрать TODO, так как не будет исключений

kanootoko commented on src/main/java/com/wine/to/up/winestyle/parser/service/controller/exception/NoEntityException.java:
```diff
@@ -0,0 +1,23 @@
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+/**
+ * Класс  ошибка, который кидается
```
Лишний пробел никуда не делся

kanootoko commented on src/main/java/com/wine/to/up/winestyle/parser/service/domain/entity/ErrorOnSaving.java:
```diff
@@ -0,0 +1,141 @@
+@Builder
+@NoArgsConstructor
+@AllArgsConstructor
+public class ErrorOnSaving {
```
Нам правда нужна такая штука в базе?

kanootoko commented on src/main/java/com/wine/to/up/winestyle/parser/service/service/implementation/controller/ParsingControllerService.java:
```diff
@@ -0,0 +1,79 @@
+                });
+                newThread.start();
+            } else {
+                throw ServiceIsBusyException.createWith(" parsing job is already running.", alcoholType);
```
Не очевидно, нужен ли здесь правда пробел, или он случайно оказался в начале строки

kanootoko commented on src/main/java/com/wine/to/up/winestyle/parser/service/service/implementation/helpers/StatusService.java:
```diff
@@ -0,0 +1,20 @@
+    private final HashMap<String, Boolean> SERVICE_BUSY_STATUS =
+            new HashMap<>(Map.of("wine", false, "sparkling", false));
+
+    public boolean statusCheck(String alcoholType) {
```
Мне кажется, метод лучше бы назвать как-нибудь вроде isParsed, а то "статус" вина, например, "ложь" - не очевидно, что значит

kanootoko commented on src/main/resources/application.properties:
```diff
@@ -1,14 +1,20 @@
 management.endpoint.prometheus.enabled=true
 
-spring.datasource.url=jdbc:postgresql://${S_POSTGRES_HOST:localhost}:5432/postgres
+spring.jpa.hibernate.ddl-auto=update
```
validate можно (и нужно) использовать, когда все создается через flyway

kanootoko commented:
Пусть пока так будет, дальше поправим
